core/corehttp/gateway_handler.go

```diff
@@ -1074,3 +1074,30 @@ func (i *gatewayHandler) setCommonHeaders(w http.ResponseWriter, r *http.Request
 
 	return nil
 }
+
+// closeConnWithError forcefully closes an HTTP/1.x connection, leading to a network
+// error on the user side. If it is not possible to forcefully close the connection,
+// we call webError which prints the error to the client, leading to a corrupt file.
+// This is a way of showing the client that there was an error while streaming the contents.
+// Currently, there are no great ways of telling the client that an error occurred while
+// streaming in HTTP.
+func closeConnWithError(w http.ResponseWriter, err error) {
+	// There are no good ways of showing an error during a stream. Therefore, we try
+	// to hijack the connection to forcefully close it, causing a network error.
+	hj, ok := w.(http.Hijacker)
+	if !ok {
+		// If we could not Hijack the connection, we write the original error. This will hopefully
+		// corrupt the generated TAR file, such that the client will receive an error unpacking.
+		webError(w, "could not build tar archive", err, http.StatusInternalServerError)
+		return
+	}
+
+	conn, _, hijackErr := hj.Hijack()
+	if hijackErr != nil {
+		// Deliberately pass the original tar error here instead of the hijacking error.
+		webError(w, "could not build tar archive", err, http.StatusInternalServerError)
+		return
+	}
+
+	conn.Close()
```

Comment:
This code doesn't work with HTTP2 afaik.

You can do `panic(http.ErrAbortHandler)` (which will be recovered) instead.
https://pkg.go.dev/net/http#ErrAbortHandler
> If ServeHTTP panics, the server (the caller of ServeHTTP) assumes that the effect of the panic was isolated to the active request. It recovers the panic, logs a stack trace to the server error log, and either closes the network connection or sends an HTTP/2 RST_STREAM, depending on the HTTP protocol. To abort a handler so the client sees an interrupted response but the server doesn't log an error, panic with the value ErrAbortHandler.

We have to be carefull to put all needed cleanup code inside `defer`s.

Reply:
> This code doesn't work with HTTP2 afaik.

Yup, I mentioned that in https://github.com/ipfs/kubo/pull/9029#issuecomment-1271201343.

> Also, I wouldn't use `panic` all the time, only if the hijack failed (mainly because it's not possible to use HTTP2 with the default Kubo install).

Definitely! I wasn't aware we could use `panic` like this here. Thanks for the pointer! 😃